sample_factory/algo/utils/action_distributions.py

```diff
@@ -81,35 +81,71 @@ def argmax_actions(distribution):
         raise NotImplementedError(f"Action distribution type {type(distribution)} does not support argmax!")
 
 
+# Retrieved from AllenNLP:
+# https://github.com/allenai/allennlp/blob/80fb6061e568cb9d6ab5d45b661e86eb61b92c82/allennlp/nn/util.py#L243
+def masked_softmax(logits, mask):
+    # To limit numerical errors from large vector elements outside the mask, we zero these out.
+    result = functional.softmax(logits * mask, dim=-1)
```

Comment:
Can you help me understand this please?

I think logits in general can be negative, or positive but close to 0, in which case multiplying them by zero does not achieve the desired effect.

I'd say we should probably use something like this instead?

```
def masked_softmax(logits, mask):
    # Mask out the invalid logits by adding a large negative number (-1e9)
    logits = logits + (mask == 0) * -1e9
    result = functional.softmax(logits, dim=-1)
    result = result * mask
    result = result / (result.sum(dim=-1, keepdim=True) + 1e-13)
    return result
```

the choice of 1e-9 is arbitrary here, but it could be something like `-max(abs(logits)) * 1e6` to make this universal

Reply:
It's got from AllenNLP including the comment so don't fully understand but as far as I investigated seems your version is safer in some cases even tho usually the results are identical in both versions 👍

Reply:
fixed